### ISSUE TITLE
Possible bug: ADS  4.0 not working with 16bit images

### ISSUE THREAD

**Issue by mathieuboudreau** [axondeepseg/axondeepseg#617]
See discussion here: https://github.com/axondeepseg/axondeepseg/discussions/614#discussioncomment-2374879

We had solved it before (https://github.com/axondeepseg/axondeepseg/issues/166 and https://github.com/axondeepseg/axondeepseg/pull/243), but the transition to IVADOMED and ADS V4 may have reintroduced the bug.

**Comment by mariehbourget** [comment]
Thanks @mathieuboudreau for opening this issue.

When running the segmentation, the files are read in ivadomed [here](https://github.com/ivadomed/ivadomed/blob/1a3d193922938dc35e9eb77d6af7fd8e6d1bacfa/ivadomed/loader/segmentation_pair.py#L311) and are supposed to be converted to 8 bits as well.
I'll take a look!

**Comment by mathieuboudreau** [comment]
Ok cool! It's possible the user misspoke, or maybe did something else when they "saved from 16 bit to 8 bit" that would have fixed their issue, but it'll be good to double check internally.

**Comment by mathieuboudreau** [comment]
Hum, looks like I uploaded the 16 bit file, but never added a test for it. My bad.

I just tried running ADS on the file, and get a similar error as the user, and it does occur when IVADOMED tries to load it "as_gray" / 8_bit, so maybe IVADOMED has a bug and not ADS?

Output:

```

(ads_venv) mathieuboudreau@Mathieus-MacBook-Pro axondeepseg % axondeepseg -t TEM -i test/__test_files__/__test_16b_file__/raw/data1/image.tif -s 0.00136 

AxonDeepSeg v.4.0.0
2022-03-16 16:00:42.644 | INFO     | ivadomed.config_manager:_display_differing_keys:153 - Adding the following keys to the configuration file
2022-03-16 16:00:42.645 | INFO     | ivadomed.config_manager:_display_differing_keys:155 - 

2022-03-16 16:00:42.645 | WARNING  | ivadomed.inference:segment_volume:402 - fname_roi has not been specified, then the entire volume is processed.
Traceback (most recent call last):
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/bin/axondeepseg", line 33, in <module>
    sys.exit(load_entry_point('AxonDeepSeg', 'console_scripts', 'axondeepseg')())
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/segment.py", line 276, in main
    segment_image(
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/segment.py", line 112, in segment_image
    axon_segmentation(path_acquisitions_folders=path_acquisition, acquisitions_filenames=[str(path_acquisition / acquisition_name)],
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 48, in axon_segmentation
    nii_lst, _ = imed_inference.segment_volume(str(path_model), input_filenames, options=options)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/inference.py", line 446, in segment_volume
    ds.load_filenames()
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/loader/mri2d_segmentation_dataset.py", line 93, in load_filenames
    roi_pair = SegmentationPair(input_filenames, roi_filename, metadata=metadata, slice_axis=self.slice_axis,
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/loader/segmentation_pair.py", line 54, in __init__
    input_img = self.read_file(input_file)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/loader/segmentation_pair.py", line 292, in read_file
    img = self.convert_file_to_nifti(filename, extension, is_gt)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/loader/segmentation_pair.py", line 314, in convert_file_to_nifti
    img = np.expand_dims(imageio.imread(filename, format='tiff-pil', as_gray=True), axis=-1).astype(np.uint8)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/imageio/core/functions.py", line 160, in imread
    return file.read(index=0, **kwargs)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/imageio/core/legacy_plugin_wrapper.py", line 134, in read
    return reader.get_data(index)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/imageio/core/format.py", line 340, in get_data
    im, meta = self._get_data(index, **kwargs)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/imageio/plugins/pillow_legacy.py", line 345, in _get_data
    im = pil_get_frame(self._im, **self._kwargs)
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/imageio/plugins/pillow_legacy.py", line 777, in pil_get_frame
    frame.shape = shape[::-1]
ValueError: cannot reshape array of size 8830976 into shape (2156,2048)

```

**Comment by mariehbourget** [comment]
I tested both PNG and TIF:

PNG with [this image](https://github.com/axondeepseg/axondeepseg/pull/243#issuecomment-531874814):
The `.astype(np.uint8)` is doing its job and convert the max value from `64764` (16 bits) to `252` (8 bits).

TIF with our [test_16b_file](https://github.com/axondeepseg/data-testing/tree/main/__test_files__/__test_16b_file__/raw/data1):
I'm unable to segment and got the same message as you.

So the bug is indeed in ivadomed with 16b TIF.

This is a dimension thing, maybe in PIL (`2156*2048 = 4415448` which is half of `8830976` from the message).

**Comment by mariehbourget** [comment]
I confirm, of all the configurations between:
- PNG or TIF
- 8 or 16 bit
- Grayscale, RGB or RGBA

The only configuration that is not working at the moment in `ivadomed` is TIF 16bit Grayscale.
I'll open a PR in ivadomed as soon as possible and get back to you.

**Comment by mariehbourget** [comment]
The bug is fixed in ivadomed 2.9.5 merged with PR #625.
I'm not closing this issue since we still need to add tests on ADS side for this.

**Comment by mariehbourget** [comment]
We can now close this issue as the tests were added in #630.
